extensions-core/orc-extensions/src/main/java/org/apache/druid/data/input/orc/OrcStructConverter.java

```diff
@@ -0,0 +1,212 @@
+      }
+    }
+    return converted;
+  }
+
+  private static Object convertPrimitive(TypeDescription fieldDescription, WritableComparable field, boolean binaryAsString)
+  {
+    /*
+        ORC TYPE    WRITABLE TYPE
+        binary      org.apache.hadoop.io.BytesWritable
+        bigint      org.apache.hadoop.io.LongWritable
+        boolean     org.apache.hadoop.io.BooleanWritable
+        char        org.apache.hadoop.io.Text
+        date        org.apache.hadoop.hive.serde2.io.DateWritable
+        decimal     org.apache.hadoop.hive.serde2.io.HiveDecimalWritable
+        double      org.apache.hadoop.io.DoubleWritable
+        float       org.apache.hadoop.io.FloatWritable
+        int         org.apache.hadoop.io.IntWritable
+        smallint    org.apache.hadoop.io.ShortWritable
+        string      org.apache.hadoop.io.Text
+        timestamp   org.apache.orc.mapred.OrcTimestamp
+        tinyint     org.apache.hadoop.io.ByteWritable
+        varchar     org.apache.hadoop.io.Text
+     */
+    switch (fieldDescription.getCategory()) {
+      case STRING:
+      case CHAR:
+      case VARCHAR:
+        return ((Text) field).toString();
+      case BOOLEAN:
+        return ((BooleanWritable) field).get();
+      case BYTE:
+        return ((ByteWritable) field).get();
+      case SHORT:
+        return ((ShortWritable) field).get();
+      case INT:
+        return ((IntWritable) field).get();
+      case LONG:
+        return ((LongWritable) field).get();
+      case FLOAT:
+        return ((FloatWritable) field).get();
+      case DOUBLE:
+        return ((DoubleWritable) field).get();
+      case DECIMAL:
+        return ((HiveDecimalWritable) field).getHiveDecimal().doubleValue();
+      case TIMESTAMP:
+        return ((OrcTimestamp) field).getTime();
+      case DATE:
+        return DateTimes.utc(((DateWritable) field).get().getTime());
+      case BINARY:
+        byte[] bytes = ((BytesWritable) field).getBytes();
+        if (binaryAsString) {
+          return StringUtils.fromUtf8(bytes);
+        } else {
+          return bytes;
+        }
+      default:
+        return null;
+    }
+  }
+
+  private boolean binaryAsString;
+
+  OrcStructConverter(boolean binaryAsString)
+  {
+    this.binaryAsString = binaryAsString;
+  }
+
+  /**
+   * Convert a orc struct field as though it were a map. Complex types will be transformed
+   * into java lists and maps when possible ({@link OrcStructConverter#convertList} and
+   * {@link OrcStructConverter#convertMap}), and
+   * primitive types will be extracted into an ingestion friendly state (e.g. 'int' and 'long'). Finally,
+   * if a field is not present, this method will return null.
+   *
+   * Note: "Union" types are not currently supported and will be returned as null
+   */
+  @Nullable
+  Object convertField(OrcStruct struct, String fieldName)
+  {
+    TypeDescription schema = struct.getSchema();
+    int fieldIndex = schema.getFieldNames().indexOf(fieldName);
```

Comment:
Might be nice to cache this in a map or something. Unless the List is specially optimized somehow, indexOf is `O(N)` in the number of fields. Shouldn't be a big deal for small numbers of fields but would get expensive for wide ORCs.

Reply:
I've added a cache for field names to index of the root `OrcStruct`, but it can't easily be used for json path expressions because the `JsonProvider` stuff doesn't have enough information to know if an `OrcStruct` it is treating as a `Map` for the transformation is the root level or not, and comparing columns to the cache seems to go full circle to where it started maybe. Better than nothing?

Reply:
Sure, it is better than nothing.